File: app/src/main/java/closer/vlllage/com/closer/handler/data/SyncHandler.java
package closer.vlllage.com.closer.handler.data;

import com.google.android.gms.maps.model.LatLng;

import java.util.List;

import closer.vlllage.com.closer.api.models.CreateResult;
import closer.vlllage.com.closer.handler.helpers.ApplicationHandler;
import closer.vlllage.com.closer.handler.helpers.ConnectionErrorHandler;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.helpers.HttpEncode;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.BaseObject;
import closer.vlllage.com.closer.store.models.Event;
import closer.vlllage.com.closer.store.models.Event_;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.GroupAction;
import closer.vlllage.com.closer.store.models.GroupAction_;
import closer.vlllage.com.closer.store.models.GroupMember;
import closer.vlllage.com.closer.store.models.GroupMember_;
import closer.vlllage.com.closer.store.models.GroupMessage;
import closer.vlllage.com.closer.store.models.GroupMessage_;
import closer.vlllage.com.closer.store.models.Group_;
import closer.vlllage.com.closer.store.models.Suggestion;
import closer.vlllage.com.closer.store.models.Suggestion_;
import io.objectbox.Property;
import io.objectbox.android.AndroidScheduler;
import io.objectbox.reactive.DataObserver;
import io.reactivex.Single;

public class SyncHandler extends PoolMember {
    public void syncAll() {
        syncAll(Suggestion.class, Suggestion_.localOnly);
        syncAll(Group.class, Group_.localOnly);
        syncAll(GroupMessage.class, GroupMessage_.localOnly);
        syncAll(Event.class, Event_.localOnly);
        syncAll(GroupAction.class, GroupAction_.localOnly);
        syncAll(GroupMember.class, GroupMember_.localOnly);
    }

    public <T extends BaseObject> void sync(T obj) {
        sync(obj, null);
    }

    public <T extends BaseObject> void sync(T obj, OnSyncResult onSyncResult) {
        send(obj, onSyncResult);
    }

    private void syncAll(Class<? extends BaseObject> clazz, Property localOnlyProperty) {
        $(StoreHandler.class).getStore().box(clazz).query()
                .equal(localOnlyProperty, true)
                .build().subscribe().single().on(AndroidScheduler.mainThread())
                .observer((DataObserver<List<? extends BaseObject>>) this::syncAll);
    }

    private void syncAll(List<? extends BaseObject> objs) {
        for (BaseObject obj : objs) {
            sync(obj);
        }
    }

    private <T extends BaseObject> void send(final T obj, OnSyncResult onSyncResult) {
        if (obj instanceof Group) {
            sendCreateGroup((Group) obj, onSyncResult);
        } else if (obj instanceof Suggestion) {
            sendCreateSuggestion((Suggestion) obj, onSyncResult);
        } else if (obj instanceof GroupMessage) {
            sendCreateGroupMessage((GroupMessage) obj, onSyncResult);
        } else if (obj instanceof Event) {
            sendCreateEvent((Event) obj, onSyncResult);
        } else if (obj instanceof GroupAction) {
            sendCreateGroupAction((GroupAction) obj, onSyncResult);
        } else if (obj instanceof GroupMember) {
            sendUpdateGroupMember((GroupMember) obj, onSyncResult);
        } else {
            throw new RuntimeException("Unknown object type for sync: " + obj);
        }
    }

    private void sendCreateGroupAction(GroupAction groupAction, OnSyncResult onSyncResult) {
        groupAction.setLocalOnly(true);
        $(StoreHandler.class).getStore().box(GroupAction.class).put(groupAction);

        $(ApplicationHandler.class).getApp().$(DisposableHandler.class).add($(ApiHandler.class).createGroupAction(
                groupAction.getGroup(),
                groupAction.getName(),
                groupAction.getIntent()
        ).subscribe(createResult -> {
            if (createResult.success) {
                groupAction.setId(createResult.id);
                groupAction.setLocalOnly(false);
                $(StoreHandler.class).getStore().box(GroupAction.class).put(groupAction);
                if (onSyncResult != null) {
                    onSyncResult.onSync(createResult.id);
                }
            }
        }, error -> $(ConnectionErrorHandler.class).notifyConnectionError()));
    }

    private void sendUpdateGroupMember(GroupMember groupMember, OnSyncResult onSyncResult) {
        groupMember.setLocalOnly(true);
        $(StoreHandler.class).getStore().box(GroupMember.class).put(groupMember);

        $(ApplicationHandler.class).getApp().$(DisposableHandler.class).add($(ApiHandler.class).updateGroupMember(
                groupMember.getGroup(),
                groupMember.isMuted(),
                groupMember.isSubscribed()
        ).subscribe(createResult -> {
            if (createResult.success) {
                groupMember.setId(createResult.id);
                groupMember.setLocalOnly(false);
                $(StoreHandler.class).getStore().box(GroupMember.class).put(groupMember);
                if (onSyncResult != null) {
                    onSyncResult.onSync(createResult.id);
                }
            }
        }, error -> $(ConnectionErrorHandler.class).notifyConnectionError()));
    }

    private void sendCreateEvent(Event event, OnSyncResult onSyncResult) {
        event.setLocalOnly(true);
        $(StoreHandler.class).getStore().box(Event.class).put(event);

        $(ApplicationHandler.class).getApp().$(DisposableHandler.class).add($(ApiHandler.class).createEvent(
                event.getName(),
                event.getAbout(),
                event.isPublic(),
                new LatLng(event.getLatitude(), event.getLongitude()),
                event.getStartsAt(),
                event.getEndsAt()
        ).subscribe(createResult -> {
            if (createResult.success) {
                event.setId(createResult.id);
                event.setLocalOnly(false);
                $(StoreHandler.class).getStore().box(Event.class).put(event);
                if (onSyncResult != null) {
                    onSyncResult.onSync(createResult.id);
                }
            }
        }, error -> $(ConnectionErrorHandler.class).notifyConnectionError()));
    }

    private void sendCreateSuggestion(Suggestion suggestion, OnSyncResult onSyncResult) {
        suggestion.setLocalOnly(true);
        $(StoreHandler.class).getStore().box(Suggestion.class).put(suggestion);

        $(ApplicationHandler.class).getApp().$(DisposableHandler.class).add($(ApiHandler.class).addSuggestion(
                suggestion.getName(),
                new LatLng(suggestion.getLatitude(), suggestion.getLongitude())
        ).subscribe(createResult -> {
                    if (createResult.success) {
                        suggestion.setId(createResult.id);
                        suggestion.setLocalOnly(false);
                        $(StoreHandler.class).getStore().box(Suggestion.class).put(suggestion);
                        if (onSyncResult != null) {
                            onSyncResult.onSync(createResult.id);
                        }
                    }
                }, error -> $(ConnectionErrorHandler.class).notifyConnectionError()));
    }

    private void sendCreateGroup(Group group, OnSyncResult onSyncResult) {
        group.setLocalOnly(true);
        $(StoreHandler.class).getStore().box(Group.class).put(group);

        Single<CreateResult> createApiRequest;

        if (group.isPhysical()) {
            createApiRequest = $(ApiHandler.class).createPhysicalGroup(new LatLng(
                    group.getLatitude(),
                    group.getLongitude()
            ));
        } else if (group.isPublic()) {
            createApiRequest = $(ApiHandler.class).createPublicGroup(group.getName(), group.getAbout(), new LatLng(
                    group.getLatitude(),
                    group.getLongitude()
            ));
        } else {
            createApiRequest = $(ApiHandler.class).createGroup(group.getName());
        }

        $(ApplicationHandler.class).getApp().$(DisposableHandler.class).add(createApiRequest.subscribe(createResult -> {
            if (createResult.success) {
                group.setId(createResult.id);
                group.setLocalOnly(false);
                $(StoreHandler.class).getStore().box(Group.class).put(group);
                if (onSyncResult != null) {
                    onSyncResult.onSync(createResult.id);
                }
            }
        }, error -> $(ConnectionErrorHandler.class).notifyConnectionError()));
    }

    private void sendCreateGroupMessage(GroupMessage groupMessage, OnSyncResult onSyncResult) {
        groupMessage.setLocalOnly(true);
        $(StoreHandler.class).getStore().box(GroupMessage.class).put(groupMessage);

        Single<CreateResult> apiCall;

        if (groupMessage.getLatitude() != null && groupMessage.getLongitude() != null) {
            apiCall = $(ApiHandler.class).sendAreaMessage(
                    new LatLng(groupMessage.getLatitude(), groupMessage.getLongitude()),
                    groupMessage.getText(),
                    $(HttpEncode.class).encode(groupMessage.getAttachment())
            );
        } else {
            apiCall = $(ApiHandler.class).sendGroupMessage(
                    groupMessage.getTo(),
                    groupMessage.getText(),
                    $(HttpEncode.class).encode(groupMessage.getAttachment())
            );
        }

        $(ApplicationHandler.class).getApp().$(DisposableHandler.class).add(apiCall.subscribe(createResult -> {
            if (createResult.success) {
                groupMessage.setId(createResult.id);
                groupMessage.setLocalOnly(false);
                $(StoreHandler.class).getStore().box(GroupMessage.class).put(groupMessage);
                if (onSyncResult != null) {
                    onSyncResult.onSync(createResult.id);
                }
            }
        }, error -> $(ConnectionErrorHandler.class).notifyConnectionError()));
    }

    public interface OnSyncResult {
        void onSync(String id);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/data/SyncHandler.kt
package closer.vlllage.com.closer.handler.data

import closer.vlllage.com.closer.api.models.CreateResult
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.*
import com.google.android.gms.maps.model.LatLng
import com.queatz.on.On
import io.objectbox.Property
import io.objectbox.android.AndroidScheduler
import io.reactivex.Single

class SyncHandler constructor(private val on: On) {
    fun syncAll() {
        syncAll(Suggestion::class.java, Suggestion_.localOnly)
        syncAll(Group::class.java, Group_.localOnly)
        syncAll(GroupMessage::class.java, GroupMessage_.localOnly)
        syncAll(Event::class.java, Event_.localOnly)
        syncAll(GroupAction::class.java, GroupAction_.localOnly)
        syncAll(GroupMember::class.java, GroupMember_.localOnly)
        syncAll(Quest::class.java, Quest_.localOnly)
        syncAll(QuestProgress::class.java, QuestProgress_.localOnly)
        syncAll(Story::class.java, Story_.localOnly)
    }

    fun <T : BaseObject> sync(obj: T, onSyncResult: OnSyncResult? = null) {
        send(obj, onSyncResult)
    }

    private fun <T : BaseObject> syncAll(clazz: Class<T>, localOnlyProperty: Property<T>) {
        on<StoreHandler>().store.box(clazz).query()
                .equal(localOnlyProperty, true)
                .build().subscribe().single().on(AndroidScheduler.mainThread())
                .observer { this.syncAll(it) }
    }

    private fun syncAll(objs: List<BaseObject>) {
        for (obj in objs) {
            sync(obj)
        }
    }

    private fun <T : BaseObject> send(obj: T, onSyncResult: OnSyncResult?) {
        when (obj) {
            is Group -> sendCreateGroup(obj, onSyncResult)
            is Suggestion -> sendCreateSuggestion(obj, onSyncResult)
            is Story -> sendCreateStory(obj, onSyncResult)
            is GroupMessage -> sendCreateGroupMessage(obj, onSyncResult)
            is Event -> sendCreateEvent(obj, onSyncResult)
            is GroupAction -> sendCreateGroupAction(obj, onSyncResult)
            is GroupMember -> sendUpdateGroupMember(obj, onSyncResult)
            is Quest -> { sendCreateQuest(obj, onSyncResult) }
            is QuestProgress -> { sendCreateQuestProgress(obj, onSyncResult) }
            else -> throw RuntimeException("Unknown object type for sync: $obj")
        }
    }

    private fun sendCreateQuest(quest: Quest, onSyncResult: OnSyncResult?) {
        quest.localOnly = true
        on<StoreHandler>().store.box(Quest::class).put(quest)

        on<ApiHandler>().createQuest(
                quest.name!!,
                quest.isPublic,
                on<LatLngStr>().to(quest.latitude!!, quest.longitude!!),
                quest.flow!!
        ).subscribe({ createResult ->
            if (createResult.success) {
                quest.id = createResult.id
                quest.localOnly = false
                on<StoreHandler>().store.box(Quest::class).put(quest)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, {
            on<ConnectionErrorHandler>().notifyConnectionError()
        }).also {
            on<ApplicationHandler>().app.on<DisposableHandler>().add(it)
        }
    }

    private fun sendCreateQuestProgress(questProgress: QuestProgress, onSyncResult: OnSyncResult?) {
        questProgress.localOnly = true
        on<StoreHandler>().store.box(QuestProgress::class).put(questProgress)

        on<ApiHandler>().createQuestProgress(
                questProgress.questId!!,
                questProgress.ofId!!,
                questProgress.isPublic,
                questProgress.active ?: true,
                questProgress.progress!!
        ).subscribe({ createResult ->
            if (createResult.success) {
                questProgress.id = createResult.id
                questProgress.localOnly = false
                on<StoreHandler>().store.box(QuestProgress::class).put(questProgress)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, {
            on<ConnectionErrorHandler>().notifyConnectionError()
        }).also {
            on<ApplicationHandler>().app.on<DisposableHandler>().add(it)
        }
    }

    private fun sendCreateGroupAction(groupAction: GroupAction, onSyncResult: OnSyncResult?) {
        groupAction.localOnly = true
        on<StoreHandler>().store.box(GroupAction::class).put(groupAction)

        on<ApplicationHandler>().app.on<DisposableHandler>().add(on<ApiHandler>().createGroupAction(
                groupAction.group!!,
                groupAction.name!!,
                groupAction.intent!!
        ).subscribe({ createResult ->
            if (createResult.success) {
                groupAction.id = createResult.id
                groupAction.localOnly = false
                on<StoreHandler>().store.box(GroupAction::class).put(groupAction)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, { error -> on<ConnectionErrorHandler>().notifyConnectionError() }))
    }

    private fun sendUpdateGroupMember(groupMember: GroupMember, onSyncResult: OnSyncResult?) {
        groupMember.localOnly = true
        on<StoreHandler>().store.box(GroupMember::class).put(groupMember)

        on<ApplicationHandler>().app.on<DisposableHandler>().add(on<ApiHandler>().updateGroupMember(
                groupMember.group!!,
                groupMember.muted,
                groupMember.subscribed
        ).subscribe({ createResult ->
            if (createResult.success) {
                groupMember.id = createResult.id
                groupMember.localOnly = false
                on<StoreHandler>().store.box(GroupMember::class).put(groupMember)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, { error -> on<ConnectionErrorHandler>().notifyConnectionError() }))
    }

    private fun sendCreateEvent(event: Event, onSyncResult: OnSyncResult?) {
        event.localOnly = true
        on<StoreHandler>().store.box(Event::class).put(event)

        on<ApplicationHandler>().app.on<DisposableHandler>().add(on<ApiHandler>().createEvent(
                event.name!!,
                event.about!!,
                event.isPublic,
                LatLng(event.latitude!!, event.longitude!!),
                event.startsAt!!,
                event.endsAt!!,
                event.allDay
        ).subscribe({ createResult ->
            if (createResult.success) {
                event.id = createResult.id
                event.localOnly = false
                on<StoreHandler>().store.box(Event::class).put(event)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, { error -> on<ConnectionErrorHandler>().notifyConnectionError() }))
    }

    private fun sendCreateSuggestion(suggestion: Suggestion, onSyncResult: OnSyncResult?) {
        suggestion.localOnly = true
        on<StoreHandler>().store.box(Suggestion::class).put(suggestion)

        on<ApplicationHandler>().app.on<DisposableHandler>().add(on<ApiHandler>().addSuggestion(
                suggestion.name!!,
                LatLng(suggestion.latitude!!, suggestion.longitude!!)
        ).subscribe({ createResult ->
            if (createResult.success) {
                suggestion.id = createResult.id
                suggestion.localOnly = false
                on<StoreHandler>().store.box(Suggestion::class).put(suggestion)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, { error -> on<ConnectionErrorHandler>().notifyConnectionError() }))
    }

    private fun sendCreateStory(story: Story, onSyncResult: OnSyncResult?) {
        story.localOnly = true
        on<StoreHandler>().store.box(Story::class).put(story)

        on<ApplicationHandler>().app.on<DisposableHandler>().add(on<ApiHandler>().addStory(
                story.text,
                story.photo,
                LatLng(story.latitude!!, story.longitude!!)
        ).subscribe({ createResult ->
            if (createResult.success) {
                story.id = createResult.id
                story.localOnly = false
                on<StoreHandler>().store.box(Story::class).put(story)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, { error -> on<ConnectionErrorHandler>().notifyConnectionError() }))
    }

    private fun sendCreateGroup(group: Group, onSyncResult: OnSyncResult?) {
        group.localOnly = true
        on<StoreHandler>().store.box(Group::class).put(group)

        val createApiRequest: Single<CreateResult> = if (group.physical) {
            on<ApiHandler>().createPhysicalGroup(group.name ?: "", group.about ?: "", group.isPublic, LatLng(
                    group.latitude!!,
                    group.longitude!!
            ), group.hub)
        } else if (group.isPublic) {
            on<ApiHandler>().createPublicGroup(group.name!!, group.about!!, LatLng(
                    group.latitude!!,
                    group.longitude!!
            ))
        } else {
            on<ApiHandler>().createGroup(group.name!!)
        }

        on<ApplicationHandler>().app.on<DisposableHandler>().add(createApiRequest.subscribe({ createResult ->
            if (createResult.success) {
                group.id = createResult.id
                group.localOnly = false
                on<StoreHandler>().store.box(Group::class).put(group)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, { error -> on<ConnectionErrorHandler>().notifyConnectionError() }))
    }

    private fun sendCreateGroupMessage(groupMessage: GroupMessage, onSyncResult: OnSyncResult?) {
        groupMessage.localOnly = true
        on<StoreHandler>().store.box(GroupMessage::class).put(groupMessage)

        val apiCall = on<ApiHandler>().sendGroupMessage(
                groupMessage.to!!,
                groupMessage.text,
                on<HttpEncode>().encode(groupMessage.attachment)
        )

        on<ApplicationHandler>().app.on<DisposableHandler>().add(apiCall.subscribe({ createResult ->
            if (createResult.success) {
                groupMessage.id = createResult.id
                groupMessage.localOnly = false
                on<StoreHandler>().store.box(GroupMessage::class).put(groupMessage)
                onSyncResult?.invoke(createResult.id!!)
            }
        }, { error -> on<ConnectionErrorHandler>().notifyConnectionError() }))
    }

}

typealias OnSyncResult = (id: String) -> Unit

